### DISCUSSION TITLE
General question about storage back-end

### DISCUSSION THREAD

**Discussion by kwoot** [GothenburgBitFactory/taskwarrior#2441] [General]
Hi,
Maybe this has been asked before (I did not find it here).
I'm no C++ expert, but does the TW code use a central storage layer for writing the data files?
If so, would it be an option to add an alternative based on SQL? SQLite would be fine.
I think it would increase the search and retrieval options immensely.
OTOH, if storage is throughout the code that idea would be a real PITA to build.
Just wondering.

**Comment by tbabej** [comment]
Moving to discussions.

**Comment by kwoot** [comment]
I took a look at the code. It seams to be very nicely structured.
It also seems that at task startup everything is read, and at the end everything is written.
That should make implementing an SQL backend relatively easy.
I guess a first step would be to generate tables based on the current data model.
And a next step to simply write everything to the database at closing time of the application.
Without a reading from the database at startup time, but having an accurate copy would already facilitate things like complex queries, a read-only web frontend and anything else people might think of.

**Comment by meonkeys** [comment]
Interesting research, @kwoot!

> I think it would increase the search and retrieval options immensely.

Do you hope to improve the speed of these operations? Taskwarrior already seems pretty fast for me. I ran `task next` on my computer a bunch of times and the average elapsed time is 50ms. `task ghistory` is slower at 150ms, but still pretty much instant compared to anything I might do in a web browser or on a mobile device.

**Comment by smemsh** [comment]
FYI, in the development branch, storage backend is being moved to a sqlite implementation in Rust.  This new layer is imported from TaskChampion project and continues development.  Sqlite will only store serialized records for now (like they are today).

It affords "real" sync and eliminates some potential transactions races, as I understand it, as well as having potential for other languages bindings to access the [serialized] task records.

**Reply by djmitche** [reply]
There's some discussion in https://github.com/GothenburgBitFactory/taskwarrior/pull/2771.

The downside is, it's just me working on it, so it's all very slow.  I'd love help :)
